fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -0,0 +1,407 @@
+
+/**
+ * Directory implementation backed by FDB which attempts to
+ * model a file system on top of FoundationDB.
+ *
+ * A few interesting details:
+ *
+ * <ul>
+ *     <li>Each segment written in Lucene once flushed is <b>immutable</b>.</li>
+ *     <li>As more data is written into Lucene, files are continually merged via a <a href="https://lucene.apache.org/core/7_6_0/core/org/apache/lucene/index/MergePolicy.html">MergePolicy</a></li>
+ * </ul>
+ *
+ * @see <a href="https://lucene.apache.org/core/7_6_0/core/org/apache/lucene/store/Directory.html">Directory</a>
+ */
+@API(API.Status.EXPERIMENTAL)
+@NotThreadSafe
+public class FDBDirectory extends Directory {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FDBDirectory.class);
+    public static final int DEFAULT_BLOCK_SIZE = 16_384;
+    private final AtomicLong nextTempFileCounter = new AtomicLong();
+    private final Transaction txn;
+    private final Subspace subspace;
+    private final Subspace metaSubspace;
+    private final Subspace dataSubspace;
+    private final byte[] sequenceSubspaceKey;
+
+    private final LockFactory lockFactory;
+    private final int blockSize;
+    private final Cache<String, FDBLuceneFileReference> fileReferenceCache;
+    private final Cache<Pair<Long, Integer>, CompletableFuture<byte[]>> blockCache;
+    private final Map<String, Long> reads;
+
+    public FDBDirectory(@Nonnull Subspace subspace, @Nonnull Transaction txn) {
+        this(subspace, txn, NoLockFactory.INSTANCE);
+    }
+
+    FDBDirectory(@Nonnull Subspace subspace, @Nonnull Transaction txn, @Nonnull LockFactory lockFactory) {
+        this(subspace, txn, lockFactory, DEFAULT_BLOCK_SIZE);
+    }
+
+    FDBDirectory(@Nonnull Subspace subspace, @Nonnull Transaction txn, @Nonnull LockFactory lockFactory, int blockSize) {
+        Verify.verify(subspace != null);
+        Verify.verify(txn != null);
+        Verify.verify(lockFactory != null);
+        this.txn = txn;
+        this.subspace = subspace;
+        final Subspace sequenceSubspace = subspace.subspace(Tuple.from(0));
+        this.sequenceSubspaceKey = sequenceSubspace.pack();
+        this.metaSubspace = subspace.subspace(Tuple.from(1));
+        this.dataSubspace = subspace.subspace(Tuple.from(2));
+        this.lockFactory = lockFactory;
+        this.blockSize = blockSize;
+        this.fileReferenceCache = CacheBuilder.newBuilder().initialCapacity(128).maximumSize(1024).recordStats().build();
+        this.blockCache = CacheBuilder.newBuilder().concurrencyLevel(16).initialCapacity(128).maximumSize(1024).recordStats().build();
+        this.reads = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Sets increment if not set yet and waits till completed to return
+     * Returns and increments the increment if its already set.
+     * @return current increment value
+     */
+    public synchronized long getIncrement() {
+        return txn.get(sequenceSubspaceKey).thenApply(
```

Comment:
Followup: Across the board in this change, we should probably be using a single specific executor, rather than the default one. And to avoid situations where code well above this calls this in an async thread and we get a deadlock, we shouldn't use the executor associated with the RecordContext.